<a href="https://colab.research.google.com/github/thegrouch4413/Predictive-Analysis/blob/main/Salaries_Exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# SF Salaries Exercise 

This Should be a quick exercise to practice your new found pandas skills! 

We will be using the [SF Salaries Dataset](https://www.kaggle.com/kaggle/sf-salaries) from Kaggle! But do not worry the dataset will be provided


**BE AWARE The tasks will get harder and harder as you go along.**

*Import pandas as pd.*

In [1]:
import pandas as pd

*Read Salaries.csv as a dataframe called sal.*

In [2]:
sal = pd.read_csv('https://raw.githubusercontent.com/glopez21/ML-Data/main/Salaries.csv')

*Check the head of the DataFrame.*

In [3]:
sal.head()

Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

*Use the .info() method to find out how many entries there are.*

In [5]:
sal.shape

(148654, 13)

In [6]:
sal.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


**What is the average BasePay ?**

In [7]:
sal['BasePay'].mean()

66325.44884050643

Tha average basepay is $66,325.45.

*What is the highest amount of OvertimePay in the dataset ?*

In [8]:
sal['OvertimePay'].max()

245131.88

The highest amount of overtime pay is $245,131.88.

*What is the job title of  JOSEPH DRISCOLL ? Note: Use all caps, otherwise you may get an answer that doesn't match up (there is also a lowercase Joseph Driscoll).*

In [19]:
sal[sal['EmployeeName']=='JOSEPH DRISCOLL']['JobTitle']

24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

Joseph Driscoll's job title is Fire Suppression Captain.

*How much does JOSEPH DRISCOLL make (including benefits)?*

In [23]:
sal[sal['EmployeeName']=='JOSEPH DRISCOLL']['TotalPayBenefits']

24    270324.91
Name: TotalPayBenefits, dtype: float64

Joseph Driscoll makes a total of $270,324.91.

*What is the name of highest paid person (including benefits)?*

In [24]:
salmax = sal['TotalPayBenefits'].idxmax()
sal.loc[salmax]['EmployeeName']

'NATHANIEL FORD'

The name of the highest pain person is Nathaniel Ford.

*What is the name of lowest paid person (including benefits)? Do you notice something strange about how much he or she is paid?*

In [26]:
salmin = sal['TotalPayBenefits'].idxmin()
sal.loc[salmin]

Id                                      148654
EmployeeName                         Joe Lopez
JobTitle            Counselor, Log Cabin Ranch
BasePay                                    0.0
OvertimePay                                0.0
OtherPay                               -618.13
Benefits                                   0.0
TotalPay                               -618.13
TotalPayBenefits                       -618.13
Year                                      2014
Notes                                      NaN
Agency                           San Francisco
Status                                     NaN
Name: 148653, dtype: object

The name of the lowest paid person is Joe Lopez. The strange thing about his pay is that it is negative. (Does he pay them?)

*What was the average (mean) BasePay of all employees per year? (2011-2014) ?*

In [27]:
sal.groupby('Year').mean()['BasePay']

Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

The average base pay for all employees for 2011 is $63,595.96.

The average base pay for all employees for 2012 is $65,436.41.

The average base pay for all employees for 2013 is $69,630.03.

The average base pay for all employees for 2014 is $66,564.42.

*How many unique job titles are there?*

In [29]:
sal['JobTitle'].nunique()

2159

There are 2,159 unique job titles in this dataset.

*What are the top 5 most common jobs?*

In [30]:
sal['JobTitle'].value_counts().head(5)

Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

The top 5 most common jobs are Transit Operator, Special Nurse, Registered Nurse, Public Service Aide and Police Officer.

*How many Job Titles were represented by only one person in 2013? (e.g. Job Titles with only one occurence in 2013?)*

In [31]:
(sal[sal['Year']==2013]['JobTitle'].value_counts()==1).sum()

202

The total number of job titles represented by one person in 2013 is 202.

*How many people have the word Chief in their job title? (This is pretty tricky)*

In [34]:
sal['JobTitle'].apply(lambda str:('chief' in str.lower()))

0         False
1         False
2         False
3         False
4          True
          ...  
148649    False
148650    False
148651    False
148652    False
148653    False
Name: JobTitle, Length: 148654, dtype: bool

In [35]:
sal['JobTitle'].apply(lambda str:('chief' in str.lower())).sum()

627

627 people have Chief included in their job title.

*Bonus: Is there a correlation between length of the Job Title string and Salary?*

In [36]:
sal["title_len"] = sal["JobTitle"].apply(lambda x : len(x))
sal[["title_len", "TotalPayBenefits"]].corr()

title_len  TotalPayBenefits
title_len          1.000000         -0.036878
TotalPayBenefits  -0.036878          1.000000

title_len  TotalPayBenefits
title_len          1.000000         -0.036878
TotalPayBenefits  -0.036878          1.000000

It seems like people with higher total pay and benefits work less as they can retire earlier. (Not sure what the correlation between the 2 are but the output of my code matches the output above.)

# Cheers!